codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/auth/AuthScheme.java

```diff
@@ -0,0 +1,222 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.ruby.codegen.auth;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.ruby.codegen.GenerationContext;
+import software.amazon.smithy.ruby.codegen.RubyCodeWriter;
+import software.amazon.smithy.ruby.codegen.config.ClientConfig;
+import software.amazon.smithy.utils.SmithyBuilder;
+
+public final class AuthScheme {
+    private final ShapeId shapeId;
+    private final String rubyAuthScheme;
+    private final String rubyIdentityType;
+    private final List<String> additionalAuthParams;
+    private final ClientConfig identityResolverConfig;
+    private final WriteAdditionalFiles writeAdditionalFiles;
+
+    private AuthScheme(Builder builder) {
+        this.shapeId = builder.shapeId;
+        this.rubyAuthScheme = builder.rubyAuthScheme;
+        this.rubyIdentityType = builder.rubyIdentityType;
+        this.additionalAuthParams = builder.additionalAuthParams;
+        this.identityResolverConfig = builder.identityResolverConfig;
+        this.writeAdditionalFiles = builder.writeAdditionalFiles;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * @return the shapeId of the auth scheme.
+     */
+    public ShapeId getShapeId() {
+        return shapeId;
+    }
+
+    /**
+     * @return the initialized ruby class of the auth scheme.
+     */
+    public String getRubyAuthScheme() {
+        return rubyAuthScheme;
+    }
+
+    /**
+     * @return the ruby identity type (class).
+     */
+    public String getRubyIdentityType() {
+        return rubyIdentityType;
+    }
+
+    /**
+     * @return additional auth params for the auth scheme.
+     */
+    public List<String> getAdditionalAuthParams() {
+        return additionalAuthParams;
+    }
+
+    /**
+     * @return the ClientConfig for the identity resolver.
+     */
+    public ClientConfig getIdentityResolverConfig() {
+        return identityResolverConfig;
+    }
+
+    /**
+     * Write additional files required by this auth scheme.
+     *
+     * @param context generation context
+     * @return List of additional files written out
+     */
+    public List<String> writeAdditionalFiles(GenerationContext context) {
+        return writeAdditionalFiles.writeAdditionalFiles(context);
+    }
+
```

Comment:
I believe we need to add an equals and hash implementation here for use in Sets/HashSet to work correctly

Reply:
What do you mean?

Reply:
Java's default for these won't work the way we want with Sets, so we need to define them and include the properties we consider important (maybe just shapeId?)

Reply:
Added equals and hashCode for shapeId

Reply:
I still don't see equals and hashCode here?

Reply:
I removed it because I stopped using Set.